[InteractionRegions] Add support for icon masking

etiennesegonzac opened the pull request:
#### 2773206a322f67c458e1f13355cb91640e8f5925
<pre>
[InteractionRegions] Add support for icon masking
<a href="https://bugs.webkit.org/show_bug.cgi?id=270982">https://bugs.webkit.org/show_bug.cgi?id=270982</a>
&lt;<a href="https://rdar.apple.com/118882637?">rdar://118882637?</a>

Reviewed by Tim Horton.

Use a shape-outside-like approach to generate InteractionRegion clip
paths for small interactive images.

* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::interactionRegionForRenderedRegion):
Add heuristics to find potential &quot;icons&quot; and generate clip paths for the
regions based on their shapes. (Using the same code as CSS `shape-outside`).

Change the way we handle simple links/buttons containers (with a single
child that also gets an Interaction Region). Instead of prioritizing the
region of the Link/Button (see EventRegion#shouldConsolidateInteractionRegion),
turn the container into a Guard and keep the child&apos;s region, which is
likely more interesting (clip path, corners etc...).

(WebCore::hasTransparentContainerStyle):
Extract the &quot;transparent container&quot; logic to a function.
(WebCore::&gt;&gt;findSVGClipElements):
Extract the code looking for an SVG shape in a function.
(WebCore::isGuardContainer):
Check to see if an element is a simple container we should turn into a
Guard to allow it&apos;s single child&apos;s to shape the Interaction Region.
(WebCore::RefPtr&lt;Image&gt;findIconImage):
Find the potential Image candidate for the raster shape generation.

* Source/WebCore/rendering/EventRegion.h:
* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegionContext::removeSuperfluousInteractionRegions):
(WebCore::EventRegionContext::uniteInteractionRegions):
Keep track of Guards rects generated for elements, and don&apos;t filter them
based on overlap.
(WebCore::guardRectForRegionBounds):
Inflate the rect for regions with complex paths that are also small.
(WebCore::EventRegionContext::shrinkWrapInteractionRegions):
White space clean-up.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
(WebKit::interactionRegionGroupNameForRegion):
Rollback a previous change for InteractionRegion layers grouping. The
current system will actually benefit from grouping regions across
layers.

* LayoutTests/interaction-region/content-hint-expected.txt:
Re-baseline with the new guard containers.
* LayoutTests/interaction-region/icon-masking-expected.txt: Added.
* LayoutTests/interaction-region/icon-masking.html: Added.
Add new test file covering icon masking.

Canonical link: <a href="https://commits.webkit.org/276139@main">https://commits.webkit.org/276139@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cd16be599a9f6777450719b860d02f816e3409d9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43797 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22846 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46234 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46441 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39895 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/46101 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26729 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20253 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/36161 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44371 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19886 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37729 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17153 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17414 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38876 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1851 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39978 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39107 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/48050 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18824 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/15473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42996 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20226 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/37999 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41701 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9757 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20423 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19850 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -205,6 +235,36 @@ static bool cachedImageIsPhoto(const CachedImage& cachedImage)
     return true;
 }
 
+static RefPtr<Image>findIconImage(const RenderObject& renderer)
```
space before the function name

hortont424 commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -205,6 +235,36 @@ static bool cachedImageIsPhoto(const CachedImage& cachedImage)
+
+static std::optional<std::pair<SVGSVGElement*, SVGGraphicsElement*>>findSVGClipElements(const RenderObject& renderer)
+{
+
```
weird newline

hortont424 commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -307,7 +367,8 @@ std::optional<InteractionRegion> interactionRegionForRenderedRegion(RenderObject
 
     float minimumPhotoArea = 200 / scale * 200 / scale;
-    if (bounds.area() > minimumPhotoArea) {
+    bool bigEnoughForPhoto = bounds.area() > minimumPhotoArea;
```
Not sure I love this name. is it really about "photo" or just "being primary content" of some sort?

hortont424 commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -205,6 +235,36 @@ static bool cachedImageIsPhoto(const CachedImage& cachedImage)
+    return nullptr;
+}
+
+static std::optional<std::pair<SVGSVGElement*, SVGGraphicsElement*>>findSVGClipElements(const RenderObject& renderer)
```
another missing space. is your editor broken? :)

hortont424 commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -329,38 +390,61 @@ std::optional<InteractionRegion> interactionRegionForRenderedRegion(RenderObject
-        auto path = svgElement.toClipPath();
+    if (iconImage && originalElement) {
+        auto imageRect = LayoutRect(rect);
+        auto shape = Shape::createRasterShape(iconImage.get(), 0, imageRect, imageRect, WritingMode::HorizontalTb, 0);
```
I guess you don't actually care about the writing mode?

etiennesegonzac replied:
nope, comment worthy?

cdumez replied:
Please don't hide smart pointers with `auto`. Should use `Ref` here.

cdumez commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -190,6 +191,35 @@ static bool shouldGetOcclusion(const RenderElement& renderer)
+            || !(style.borderTopWidth() && style.borderRightWidth() && style.borderBottomWidth() && style.borderLeftWidth()));
+}
+
+static bool isGuardContainer(Element& element)
```
nit: Could this be a `const Element&`

cdumez commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -205,6 +235,35 @@ static bool cachedImageIsPhoto(const CachedImage& cachedImage)
+    return nullptr;
+}
+
+static std::optional<std::pair<SVGSVGElement*, SVGGraphicsElement*>> findSVGClipElements(const RenderObject& renderer)
```
This should likely not be returning a pair of raw pointers. It is risky security-wise. We likely want to have RefPtrs or WeakPtrs instead of raw pointers since the caller cannot easily assign the raw pointers to WeakPtr/RefPtr.

cdumez replied:
Looking at the rest of the code, RefPtr seems appropriate.

etiennesegonzac replied:
Thanks for the ... pointer :)

cdumez commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -329,38 +389,61 @@ std::optional<InteractionRegion> interactionRegionForRenderedRegion(RenderObject
         return std::nullopt;
 
+    RefPtr<Image> iconImage;
+    std::optional<std::pair<SVGSVGElement*, SVGGraphicsElement*>> svgClipElements;
```
Storing raw pointers here is risky.

cdumez commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -329,38 +389,61 @@ std::optional<InteractionRegion> interactionRegionForRenderedRegion(RenderObject
-        auto& svgElement = renderShape->graphicsElement();
-        auto path = svgElement.toClipPath();
+    if (iconImage && originalElement) {
+        auto imageRect = LayoutRect(rect);
```
`LayoutRect imageRect(rect);` is more concise and seems clearer.

etiennesegonzac committed:
[InteractionRegions] Add support for icon masking
https://bugs.webkit.org/show_bug.cgi?id=270982
<rdar://118882637?

Reviewed by Tim Horton.

Use a shape-outside-like approach to generate InteractionRegion clip
paths for small interactive images.

* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::interactionRegionForRenderedRegion):
Add heuristics to find potential "icons" and generate clip paths for the
regions based on their shapes. (Using the same code as CSS `shape-outside`).

Change the way we handle simple links/buttons containers (with a single
child that also gets an Interaction Region). Instead of prioritizing the
region of the Link/Button (see EventRegion#shouldConsolidateInteractionRegion),
turn the container into a Guard and keep the child's region, which is
likely more interesting (clip path, corners etc...).

(WebCore::hasTransparentContainerStyle):
Extract the "transparent container" logic to a function.
(WebCore::>>findSVGClipElements):
Extract the code looking for an SVG shape in a function.
(WebCore::isGuardContainer):
Check to see if an element is a simple container we should turn into a
Guard to allow it's single child's to shape the Interaction Region.
(WebCore::RefPtr<Image>findIconImage):
Find the potential Image candidate for the raster shape generation.

* Source/WebCore/rendering/EventRegion.h:
* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegionContext::removeSuperfluousInteractionRegions):
(WebCore::EventRegionContext::uniteInteractionRegions):
Keep track of Guards rects generated for elements, and don't filter them
based on overlap.
(WebCore::guardRectForRegionBounds):
Inflate the rect for regions with complex paths that are also small.
(WebCore::EventRegionContext::shrinkWrapInteractionRegions):
White space clean-up.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
(WebKit::interactionRegionGroupNameForRegion):
Rollback a previous change for InteractionRegion layers grouping. The
current system will actually benefit from grouping regions across
layers.

* LayoutTests/interaction-region/content-hint-expected.txt:
Re-baseline with the new guard containers.
* LayoutTests/interaction-region/icon-masking-expected.txt: Added.
* LayoutTests/interaction-region/icon-masking.html: Added.
Add new test file covering icon masking.

Canonical link: https://commits.webkit.org/276139@main
